### PR TITLE
Escape Go templates

### PR DESCRIPTION
**Pull request by divolgin**

### CODE REVIEW

File: pkg/socketservice/helm.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+// When saving templated file back into a chart, we need to escape Go templates so second Helm pass would ignore them.
+// These are appication templates that maybe used in application config files and Helm should ignore them.
+// For example, origianl chart has this:
+//		"legendFormat": "{{`{{`}} value {{`}}`}}",
+// Rendered chart becomes:
+//		"legendFormat": "{{ value }}",
+// Repackaged chart should have this:
+//		"legendFormat": "{{`{{`}} value {{`}}`}}",
+func escapeGoTemplates(content []byte) []byte {
```
Comment (divolgin) [inline pkg/socketservice/helm.go:32]: This is new

File: pkg/socketservice/helm.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+	return archive, nil
+}
+
+func saveHelmFile(rootDir string, relDir string, filename string, content []byte) error {
+	// We only get CRDs and templates YAML after kustomization
+	destDir := filepath.Join(rootDir, relDir)
+	if filepath.Base(relDir) != "crds" {
+		destDir = filepath.Join(destDir, "templates")
+		content = escapeGoTemplates(content)
```
Comment (divolgin) [inline pkg/socketservice/helm.go:135]: this line is new

File: pkg/socketservice/helm.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+	fileContent, err := ioutil.ReadFile(filepath.Join(srcRootDir, relPath, filename))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return errors.Wrap(err, "failed to read file")
+	}
+
+	dstDir := filepath.Join(dstRootDir, relPath)
+	if err := os.MkdirAll(dstDir, 0744); err != nil {
```
Comment (emosbaugh) [inline pkg/socketservice/helm.go:161]: 755 or 700?

File: pkg/socketservice/helm.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+func saveHelmFile(rootDir string, relDir string, filename string, content []byte) error {
+	// We only get CRDs and templates YAML after kustomization
+	destDir := filepath.Join(rootDir, relDir)
+	if filepath.Base(relDir) != "crds" {
+		destDir = filepath.Join(destDir, "templates")
+		content = escapeGoTemplates(content)
+	}
+
+	if err := os.MkdirAll(destDir, 0744); err != nil {
```
Comment (emosbaugh) [inline pkg/socketservice/helm.go:138]: 755 or 700?

File: pkg/socketservice/helm.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	metadataFiles := []string{"Chart.yaml", "Chart.lock"}
+
+	err = filepath.Walk(archiveChartDir,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			relPath, err := filepath.Rel(archiveChartDir, filepath.Dir(path))
+			if err != nil {
+				return err
```
Comment (emosbaugh) [inline pkg/socketservice/helm.go:79]: wrap

File: pkg/socketservice/helm.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+	desrChartsDir := filepath.Join(exportChartPath, "charts")
+	if _, err := os.Stat(desrChartsDir); os.IsNotExist(err) {
+		if err := os.MkdirAll(desrChartsDir, 0744); err != nil {
+			return nil, errors.Wrap(err, "failed to mkdir for archive chart")
+		}
+	}
+
+	sourceChartsDir := filepath.Join(deployedVersionArchive, "base", "charts")
+	metadataFiles := []string{"Chart.yaml", "Chart.lock"}
```
Comment (emosbaugh) [inline pkg/socketservice/helm.go:70]: i wonder if we need any other files

https://github.com/helm/helm/blob/21251378b432fb629cb1a49af3892e20c1ec9624/pkg/chart/loader/load.go#L95-L155

not really for this pr but something to think about.

**Review by emosbaugh — APPROVED** [review]

File: pkg/socketservice/helm.go (hunk 6 above)
Comment (divolgin) [reply pkg/socketservice/helm.go:70]: interesting.  the requirements files might be of interest.  i don't know what's in those.  but values should be used on initial render and can be discarded.

**Pull request merged**